orm_lib/inc/drogon/orm/SqlBinder.h

```diff
@@ -373,21 +371,6 @@ class DROGON_EXPORT SqlBinder : public trantor::NonCopyable
         std::shared_ptr<void> obj = std::make_shared<ParaType>(parameter);
         if (type_ == ClientType::PostgreSQL)
         {
-#if __cplusplus >= 201703L || (defined _MSC_VER && _MSC_VER > 1900)
-            const size_t size = sizeof(T);
-            if constexpr (size == 2)
-            {
-                *std::static_pointer_cast<uint16_t>(obj) = htons(parameter);
-            }
-            else if constexpr (size == 4)
-            {
-                *std::static_pointer_cast<uint32_t>(obj) = htonl(parameter);
-            }
-            else if constexpr (size == 8)
-            {
-                *std::static_pointer_cast<uint64_t>(obj) = htonll(parameter);
-            }
-#else
```

Comment:
How about:
```c++
            if constexpr (sizeof(T) == 2)
            {
                *std::static_pointer_cast<uint16_t>(obj) = htons(parameter);
            }
            else if constexpr (sizeof(T) == 4)
            {
                *std::static_pointer_cast<uint32_t>(obj) = htonl(parameter);
            }
            else if constexpr (sizeof(T) == 8)
            {
                *std::static_pointer_cast<uint64_t>(obj) = htonll(parameter);
            }
```

Reply:
The statements (such as `*std::static_pointer_cast<uint16_t>(obj) = htons(parameter);`) do not depend on template arguments, so they are completely the same as normal `if` branches.

The following code includes a statement that depends on template arguments:
(https://github.com/EzoeRyou/cpp17book/blob/master/028-cpp17-core-constexpr-if.md#%E8%B6%85%E4%B8%8A%E7%B4%9A%E8%80%85%E5%90%91%E3%81%91%E8%A7%A3%E8%AA%AC)

```cpp
template  < typename ... >
bool false_v = false ;

template < typename T >
void f()
{
    if constexpr ( E1 )
        if constexpr ( E2 )
            if constexpr ( E3 )
            {
                static_assert( false_v<T> ) ;
            }
}
```

`constexpr if` is used for a conditional template instantiation, not a conditional compilation. If you would avoid the compilation, you can use preprocessor directives, but I do not recommend using them.

```cpp
#if sizeof(T) == 2
*std::static_pointer_cast<uint16_t>(obj) = htons(parameter);
#elif sizeof(T) == 4
*std::static_pointer_cast<uint32_t>(obj) = htonl(parameter);
...
```

Reply:
> The statements (such as `*std::static_pointer_cast<uint16_t>(obj) = htons(parameter);`) do not depend on template arguments, so they are completely the same as normal `if` branches.
> 
> The following code includes a statement that depends on template arguments: (https://github.com/EzoeRyou/cpp17book/blob/master/028-cpp17-core-constexpr-if.md#%E8%B6%85%E4%B8%8A%E7%B4%9A%E8%80%85%E5%90%91%E3%81%91%E8%A7%A3%E8%AA%AC)
> 
> ```c++
> template  < typename ... >
> bool false_v = false ;
> 
> template < typename T >
> void f()
> {
>     if constexpr ( E1 )
>         if constexpr ( E2 )
>             if constexpr ( E3 )
>             {
>                 static_assert( false_v<T> ) ;
>             }
> }
> ```
> 
> `constexpr if` is used for a conditional template instantiation, not a conditional compilation. If you would avoid the compilation, you can use preprocessor directives, but I do not recommend using them.
> 
> ```c++
> #if sizeof(T) == 2
> *std::static_pointer_cast<uint16_t>(obj) = htons(parameter);
> #elif sizeof(T) == 4
> *std::static_pointer_cast<uint32_t>(obj) = htonl(parameter);
> ...
> ```

@ken-matsui thanks so much for your patch and reply.
I think the `if constexpr` and the preprocessor you mentioned all avoid the checking sizeof(T) in runtime, this can make some little performance advantage.

Reply:
@an-tao 
Thank you, too :) For that use-case, we can use a constexpr variable:

```cpp
constexpr size_t size = sizeof(T);
```

After that, we can use the variable in runtime:

```cpp
switch (size)
```

WDYT?

Reply:
IMHO, `if constexpr` or `#if ...` can also avoid the switch branches in runtime.